beets/dbcore/db.py

```diff
@@ -670,26 +654,22 @@ class Results:
     constructs LibModel objects that reflect database rows.
     """
 
-    def __init__(self, model_class, rows, db, flex_rows,
-                 query=None, sort=None):
+    def __init__(self, model_class, rows, db, flex_rows, sort=None):
```

Comment:
This is very much a strategic thing about backwards compatibility… but what would you think about leaving the "slow query" infrastructure in place here as a kind of fallback? In particular, I think it is important to have in case plugins provide slow queries that cannot be implemented in the new style. We can leave this functionality here and just gradually reduce our reliance on it—and someday in the future remove it after it's truly been proven worthless.

This kind of fallback could also support an incremental approach to your current work. For instance, if querying multiple flexattrs turns out to be kinda complicated, we can leave those as "slow" queries and they will at least keep working (and their performance will be no worse than before). Then we can address them in a future iteration.

Reply:
That's a very good point. Indeed, we shouldn't make a hard switch - hopefully this might also take care of the `lslimit` testing failures, so that this PR can be merged in smoothly before we get back to #4731. Still getting used to maintaining public APIs :)

> I think it is important to have in case plugins provide slow queries that cannot be implemented in the new style.

I reckon it should eventually be replaced entirely - I think it will be easier to maintain it when there's just a single way of doing this?

I wouldn't be against having this as a constraint for plugin developers - they may try and implement the filter in SQL; if their query is really only possible in Python, we could provide them with the `add_db_function` (or equivalent) [that we discussed the other day](https://github.com/beetbox/beets/tree/generic-db-functions) as a fallback? Of course, we'd have to figure out the design before we replace it completely. 

This should ensure that all queries are fast in the future, methinks 🙂 What do you think about deprecating this logic, and making this clear in the changelog?